3rdParty/fuerte/src/AsioSockets.h

```diff
@@ -137,12 +144,45 @@ struct Socket<fuerte::SocketType::Ssl> {
     });
   }
   
-  void shutdown() {
+  void shutdown(std::shared_ptr<fuerte::Connection> const& connection) {
+    // The connection object is needed to hand it to the completion handlers
+    // to ensure that the connection object is still alive as long as
+    // asynchronous operations are ongoing. It needs to be a connection
+    // object using this socket for communications! This will in turn
+    // keep ourselves alive for long enough! Note that both completion
+    // handlers below intentionally copy the connection shared_ptr!
+    if (socket.lowest_layer().is_open()) {
+      asio_ns::error_code ec; // ignored
+      socket.lowest_layer().cancel(ec);
+      // Set a timeout, since the shutdown could block forever, if we
+      // reach the timeout, we simply give up on TLS and close the TCP
+      // layer. Note the cleanupDone is necessary, since it is possible
+      // that both the completion handler of the timeout and the one
+      // of the async_shutdown are called without error!
+      timer.expires_from_now(std::chrono::seconds(3));
+      timer.async_wait([connection, this](asio_ns::error_code ec) {
+        if (!cleanupDone && !ec) {
```

Comment:
Is it possible that the callbacks for `async_wait` and `async_shutdown` are called concurrently by different IO threads? In that case we would have a race here since `cleanupDone` is only set after the call to `shutdownTcp`.

Reply:
As far as I know we only ever put one thread in an io_context, so that this cannot happen.

Reply:
Well, currently yes, we only put one in. But we just made it so that you can use more: https://github.com/arangodb/arangodb/pull/13049

Reply:
@danielhlarkin Yes, we may have multiple IO threads, but each thread has its own io_context. And since the timer and socket are both on the same context, the callbacks cannot run concurrently. So this should be fine.

Reply:
Ah, okay, I hadn't dug into the details there. Good.